NDTensors/ext/NDTensorsMetalExt/adapt.jl

```diff
@@ -6,6 +6,7 @@ end
 
 # More general than the version in Metal.jl
 function Adapt.adapt_storage(arraytype::Type{<:MtlArray}, xs::AbstractArray)
-  arraytype_specified = specify_parameters(arraytype, get_parameters(xs))
+  params = get_parameters(xs)
+  arraytype_specified = specify_parameters(arraytype, params[1], params[2], params[3:end])
```

Comment:
I'm confused by this rewrite. What's the advantage of the new way you wrote it?

Also, there seems to be a few issues here:
1. `get_parameters` is the syntax from `SetParameters`, I think you should be using `type_parameters`.
2. It seems like this isn't general enough. For example, `xs` may be an array type that doesn't have type parameters that match those of `MtlArray` (for example, maybe it is a CPU array so doesn't have a storage mode/buffer type, maybe it is a wrapper array that doesn't have an `ndims` or even an `eltype` field, etc. One way I could imagine handling that is a pattern like this:
```julia
param_names = (eltype, ndims, storagemode)
params = type_parameters(xs, param_names)
arraytype_specified = specify_parameters(arraytype, param_names, params)
```
This may not be quite right, since we would need to check if `xs` in fact has a storage mode, and if not, we should use `UnspecifiedTypeParameter`. I think right now `type_parameters` will error if there is no `storagemode` defined. I forget if we have a more lenient alternative to `type_parameters` that either outputs the type parameters or `UnspecifiedTypeParameter`, but if not we should add that `TypeParameterAccessors`.

Reply:
Another option would be to define `parameter_names(xs)` which would output just the parameter names that are valid for `xs`, for example it could output `(eltype, ndims)` if `xs isa Array`. Then it could be written:
```julia
param_names = parameter_names(xs)
params = type_parameters(xs, param_names)
arraytype_specified = specify_parameters(arraytype, param_names, params)
```
However, that may not be the best approach. For example, `xs` may be a matrix wrapper type like `Hermitian` which doesn't have an `ndims` parameter field, but `ndims` is implicitly `2`:
```julia
julia> using LinearAlgebra

julia> typeof(Hermitian(randn(2, 2)))
Hermitian{Float64, Matrix{Float64}}

julia> ndims(typeof(Hermitian(randn(2, 2))))
2
```
so that wouldn't quite work, since `parameter_names(xs::Hermitian)` would be `(eltype, parenttype)`, and not have an `ndims` name. But that kind of idea could be used to make the code from the previous post a bit more general:
```julia
param_names = parameter_names(arraytype)
params = type_parameters(xs, param_names)
arraytype_specified = specify_parameters(arraytype, param_names, params)
```

Reply:
Anyway, maybe we don't need to address this issue in full generality in this PR, and just focus on making the code work as well as it did before the PR, and fix this in a future PR.

Reply:
Another thing, isn't this definition type piracy?

Reply:
Thank you for your messages. Currently in the `main` branch of ITensors  this function`NDTensor.mtl` is **broken**. This is an effort to patch the problem without significantly modifying how the code is written because in subsequent PRs I will be modifying the code in general.
In this PR I am not migrating the code to use `TypeParameterAccessors` beyond `unwrap_array_type` and `parenttype` to limit the scope. Because of this these changes are simply a short term patch to get the code working again. In the Next PR I will be changing this code since I will be removing `SetParameters`and I can integrate the changes in `TypeParameterAccessors` to make this code more generic and robust. But again for now this is a short term change just to get the metal code working enough to run the unittests.

Reply:
I see, thanks for the explanation. Makes sense to do this in stages as you outline. Though I'm still confused by why the change you made is better than the previous version.

Reply:
The issue that I was seeing is that the function 
```
arraytype_specified = specify_parameters(arraytype, get_parameters(xs))
```
Which ultimately calls this 
```
function set_parameters(
  set_parameter_function::Function,
  type::Type,
  position::Position,
  parameter1,
  parameters_tail...,
)
  new_type = set_parameter_function(type, position, parameter1)
  new_position = Position(get_parameter(position) + 1)
  return set_parameters(set_parameter_function, new_type, new_position, parameters_tail...)
end
```
The problem was `get_parameters(...)` was returning a tuple so for some reason `parameter1` was being interpreted as a tuple
```
using NDTensors.SetParameters
f(param1, param2...) = println("$(param1)")
f(get_parameters(Vector{Float32})) # = (Float32, 1)
```
and this was incorrect. This shouldn't be a problem in our new system.

Reply:
I see, so can you just write it as:
```julia
  params = get_parameters(xs)
  arraytype_specified = specify_parameters(arraytype, params...)
```
?

Reply:
Yes this works so I updated the code to match your suggestion. Thanks!